pkg/controller/controller_ref_manager.go

```diff
@@ -346,6 +346,7 @@ func (m *ReplicaSetControllerRefManager) AdoptReplicaSet(rs *extensions.ReplicaS
 	if err := m.canAdopt(); err != nil {
 		return fmt.Errorf("can't adopt ReplicaSet %v/%v (%v): %v", rs.Namespace, rs.Name, rs.UID, err)
 	}
+	rs.Finalizers = append(rs.Finalizers, metav1.FinalizerDeleteDependents)
```

Comment:
We intend to set the Replicaset.ObjectMeta.Finalizer to DeletingDependents when deployment create/adopt the replicaset, otherwise we hit #44046.

@krmayankk we shouldn't change the object we get from the informer. We should send a patch to the apiserver here.